drivers/frequency/adf5355/adf5355.c

```diff
@@ -0,0 +1,523 @@
+	} else {
+		cp_bleed = DIV_ROUND_UP(400 * dev->cp_ua, dev->integer * 375);
+	}
+
+	cp_bleed = clamp(cp_bleed, 1U, 255U);
+
+	dev->regs[ADF5355_REG(0)] = ADF5355_REG0_INT(dev->integer) |
+				    ADF5355_REG0_PRESCALER(prescaler) |
+				    ADF5355_REG0_AUTOCAL(1);
+
+	dev->regs[ADF5355_REG(1)] = ADF5355_REG1_FRACT(dev->fract1);
+
+	dev->regs[ADF5355_REG(2)] = ADF5355_REG2_MOD2(dev->mod2) |
+				    ADF5355_REG2_FRAC2(dev->fract2);
+
+	if (dev->dev_id == ADF5356)
+		dev->regs[ADF5355_REG(13)] = ADF5356_REG13_MOD2_MSB(dev->mod2 >> 14) |
+					     ADF5356_REG13_FRAC2_MSB(dev->fract2 >> 14);
+
+	dev->regs[ADF5355_REG(6)] = ADF5355_REG6_OUTPUT_PWR(dev->outa_power) |
+				    ADF5355_REG6_RF_OUT_EN(dev->outa_en) |
+				    (dev->dev_id == ADF5355 ? ADF5355_REG6_RF_OUTB_EN(!dev->outb_en) :
+				     ADF4355_REG6_OUTPUTB_PWR(dev->outb_power) |
+				     ADF4355_REG6_RF_OUTB_EN(dev->outb_en)) |
+				    ADF5355_REG6_MUTE_TILL_LOCK_EN(dev->mute_till_lock_en) |
+				    ADF5355_REG6_CP_BLEED_CURR(cp_bleed) |
+				    ADF5355_REG6_RF_DIV_SEL(dev->rf_div_sel) |
+				    ADF5355_REG6_FEEDBACK_FUND(1) |
+				    ADF5355_REG6_NEG_BLEED_EN(cp_neg_bleed_en) |
+				    ADF5355_REG6_GATED_BLEED_EN(dev->cp_gated_bleed_en) |
+				    ADF5356_REG6_BLEED_POLARITY(dev->dev_id == ADF5356 ?
+						    dev->cp_bleed_current_polarity_en : 0) |
+				    ADF5355_REG6_DEFAULT;
+
+	dev->freq_req = freq;
+
+	return adf5355_reg_config(dev, dev->all_synced);
+}
+
+/**
+ * Get the output frequency of one channel.
+ * @param dev - The device structure.
+ * @param channel - The selected channel.
+ * @return The ouput frequency.
+ */
+static uint64_t adf5355_pll_fract_n_get_rate(struct adf5355_dev *dev,
+		uint32_t channel)
+{
+	uint64_t val, tmp;
+
+	val = (((uint64_t)dev->integer * ADF5355_MODULUS1) + dev->fract1) * dev->fpfd;
+	tmp = (uint64_t)dev->fract2 * dev->fpfd;
+	do_div(&tmp, dev->mod2);
+	val += tmp + ADF5355_MODULUS1 / 2;
+
+	do_div(&val, ADF5355_MODULUS1 *
+	       (1 << (channel == 1 ? 0 : dev->rf_div_sel)));
+
+	if (channel == 1)
+		val <<= 1;
+
+	return val;
+}
+
+/**
+ * Recalculate rate corresponding to a channel.
+ * @param dev - The device structure.
+ * @param chan - Channel number.
+ * @param rate - Channel rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf5355_clk_recalc_rate(struct adf5355_dev *dev, uint32_t chan,
+				uint64_t *rate)
+{
+	if (chan > dev->num_channels)
+		return FAILURE;
+
+	*rate = adf5355_pll_fract_n_get_rate(dev, chan);
+
+	return SUCCESS;
+}
+
+/**
+ * Set channel rate.
+ * @param dev - The device structure.
+ * @param chan - Channel number.
+ * @param rate - Channel rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf5355_clk_set_rate(struct adf5355_dev *dev, uint32_t chan,
+			     uint64_t rate)
+{
+	if (chan >= dev->num_channels)
+		return FAILURE;
+
+	return adf5355_set_freq(dev, rate, chan);
```

Comment:
Probably the logic inside `adf5355_set_freq` should be moved inside `adf5355_clk_round_rate`, and only some registers settings should take place in `adf5355_set_freq` function, otherwise the output `rounded_rate` of function `adf5355_clk_round_rate` and the actual value set with `adf5355_clk_set_rate` might be different.
This functions should be used together like in this case:
  https://github.com/analogdevicesinc/no-OS/blob/1365adbebfef34edb32a5ce89dce603f1bb93451/projects/adrv9009/src/app/app_clocking.c#L459
`clk_round_rate` called first to determine valid rate
https://github.com/analogdevicesinc/no-OS/blob/1365adbebfef34edb32a5ce89dce603f1bb93451/projects/adrv9009/src/app/app_clocking.c#L487
`clk_set_rate` second to set the determined valid rate

Reply:
This looks like a valid approach to me.
The only isssue I see is the alignment with other clock chip drivers on no-OS/linux which use more or less the current approach:
https://github.com/analogdevicesinc/no-OS/tree/master/drivers/frequency/adf4371
https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/frequency/adf5355.c
Even for hmc7044, you showed an example of usage of the driver, but if we have a look at the driver implementation in the no-OS repository, the frequency parameters computation is done mainly in the `setup` function